File: game/src/main/java/me/ryleykimmel/brandywine/game/message/listener/CommandMessageListener.java
package me.ryleykimmel.brandywine.game.message.listener;

import java.util.Arrays;
import me.ryleykimmel.brandywine.common.Strings;
import me.ryleykimmel.brandywine.game.command.CommandArguments;
import me.ryleykimmel.brandywine.game.command.CommandEvent;
import me.ryleykimmel.brandywine.game.message.CommandMessage;
import me.ryleykimmel.brandywine.game.model.World;
import me.ryleykimmel.brandywine.game.model.player.Player;

/**
 * Listener for the {@link CommandMessage}.
 */
public final class CommandMessageListener extends GameMessageListener<CommandMessage> {

  /**
   * Constructs a new {@link CommandMessageListener}.
   *
   * @param world The World listening to this {@link CommandMessage}.
   */
  public CommandMessageListener(World world) {
    super(world);
  }

  @Override
  public void handle(Player player, CommandMessage message) {
    String[] components = message.getCommand().split(" ");
    String name = components[0];

    String[] filtered = Arrays.copyOfRange(components, 1, components.length);
    String[] arguments = Strings.split(String.join(" ", filtered), '"');

    world.notify(new CommandEvent(player, name, new CommandArguments(arguments)));
  }

}

File: game/src/main/kotlin/plugin/message/CommandMessageListener.kt
package plugin.message

import me.ryleykimmel.brandywine.common.Strings
import me.ryleykimmel.brandywine.game.command.CommandArguments
import me.ryleykimmel.brandywine.game.command.CommandEvent
import me.ryleykimmel.brandywine.game.model.player.Player
import me.ryleykimmel.brandywine.game.msg.CommandMessage
import plugin.world
import java.util.*

class CommandMessageListener : MessageListener<Player, CommandMessage> {

    override fun handle(source: Player, message: CommandMessage) {
        val components = message.command.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val name = components[0]

        val filtered = Arrays.copyOfRange(components, 1, components.size)
        val arguments = Strings.split(filtered.joinToString(" "), '"')

        world.notify(CommandEvent(source, name, CommandArguments(arguments)))
    }

}
